### DISCUSSION TITLE
Add Notes to dataset export

### DISCUSSION THREAD

**Discussion by julianemoellmann** [asreview/asreview#1220] [Ideas]
First of all, thanks for creating asreview. It is a really great tool and the usability is awesome.

In my review, I created notes to justify and trace my decision later on, but I realized that the notes are not exported when I export the database. My co-author would instead review my choices in excel than in asreview. Is it possible to also export the note related to the dataset item?

**Comment by Rensvandeschoot** [comment]
There are two options:

1. Ask your colleague to install ASReview and send this person your [exported ](https://asreview.readthedocs.io/en/latest/manage.html#export-project)`.asreview` file. Your colleague should then [import ](https://asreview.readthedocs.io/en/latest/manage.html#import-project) the file and open the history panel. Your colleague can now use the filter to select the records containing notes. 

2. You can add the notes from the `.asreview` project file to the `. csv` file via the[ASReview Python API](https://asreview.readthedocs.io/en/latest/example_api_asreview_file.html).

![image](https://user-images.githubusercontent.com/36502709/189834196-51c2c55a-72bb-43cd-97bd-2c2d6c77b672.png)

**Comment by julianemoellmann** [comment]
Hey Rens,
thanks for your fast answer. 

Unfortunately, Option 1 is not feasible since my colleague cannot search for specific references in the history panel. If he/she wants to access the note of a specific reference because, for example, I noted there why I think a reference is relevant or irrelevant, he/she cannot do this. And I cannot really ask them to scroll through thousands of references.

**Reply by Rensvandeschoot** [reply]
I understand! So, option 2 it is :-)

**Comment by elisebvd** [comment]
I have the same question. Unfortunately but I'm not a python user, so I have no clue how to do this. Would you be able to provide a step-by-step guide for dummies? That would be great. And it would be even greater if this could be build into ASReview lab, so it would just always include the notes in the export.

**Reply by rohitgarud** [reply]
Hi.. were you able to figure it out? Were you able to use Python API to get notes? Thank you.

**Reply by elisebvd** [reply]
Hi, yes, in the end I managed to get the notes with Python API instructions, but it was quite a puzzle.

**Comment by Rensvandeschoot** [comment]
Great to hear!!! If you can share the steps you have taken, we can create this tutorial together for on the blog: https://asreview.nl/blog/

**Reply by rohitgarud** [reply]
Actually, after going through same issue, I have created extension for ASReview which can be installed using pip. @Rensvandeschoot sir, please review it at [asreview-notes-export](https://github.com/rohitgarud/asreview-notes-export) give me suggestions to improve it. Thank you very much.

**Comment by j0sien** [comment]
Great @rohitgarud  it worked out for me for the first screening round using the default settings.
However, when trying to export after the second screening round (using the exported decisions with notes from the first round as import), I get error messages:

C:\Users\HU\Documents>asreview notes_export ScreeningRound2.asreview
Traceback (most recent call last):
  File "C:\Python310\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Python310\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Python310\Scripts\asreview.exe\__main__.py", line 7, in <module>
  File "C:\Users\HU\AppData\Roaming\Python\Python310\site-packages\asreview\__main__.py", line 78, in main
    entry.load()().execute(sys.argv[2:])
  File "C:\Python310\lib\site-packages\asreviewcontrib\notes_export\entrypoint.py", line 63, in execute
    export_notes(
  File "C:\Python310\lib\site-packages\asreviewcontrib\notes_export\notes_export.py", line 36, in export_notes
    dataset_with_results = dataset.df.join(df.set_index('record_id')[['labeling_order','label','exported_notes']], on='record_id')
  File "C:\Python310\lib\site-packages\pandas\core\frame.py", line 9260, in join
    return self._join_compat(
  File "C:\Python310\lib\site-packages\pandas\core\frame.py", line 9291, in _join_compat
    return merge(
  File "C:\Python310\lib\site-packages\pandas\core\reshape\merge.py", line 122, in merge
    return op.get_result()
  File "C:\Python310\lib\site-packages\pandas\core\reshape\merge.py", line 718, in get_result
    llabels, rlabels = _items_overlap_with_suffix(
  File "C:\Python310\lib\site-packages\pandas\core\reshape\merge.py", line 2317, in _items_overlap_with_suffix
    raise ValueError(f"columns overlap but no suffix specified: {to_rename}")
ValueError: columns overlap but no suffix specified: Index(['labeling_order', 'exported_notes'], dtype='object')

C:\Users\HU\Documents>asreview notes_export ScreeningRound2.asreview ScreeningRound2.csv
Traceback (most recent call last):
  File "C:\Python310\lib\runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Python310\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Python310\Scripts\asreview.exe\__main__.py", line 7, in <module>
  File "C:\Users\HU\AppData\Roaming\Python\Python310\site-packages\asreview\__main__.py", line 78, in main
    entry.load()().execute(sys.argv[2:])
  File "C:\Python310\lib\site-packages\asreviewcontrib\notes_export\entrypoint.py", line 47, in execute
    raise ValueError("Exporting notes from multiple project files"
ValueError: Exporting notes from multiple project files via the CLI is not supported yet.

**Reply by rohitgarud** [reply]
Hi.. @j0sien  thank you for your feedback. I think I know what is causing the error. I will make corrections and update the version of the extension on pypi.. In the meantime, can you please raise the issue in the GitHub repository of the extension at [asreview-notes-export](https://github.com/rohitgarud/asreview-notes-export) so that we can discuss it further there and track the issue. Thank you..

**Reply by rohitgarud** [reply]
Also, there is one small correction in the command you used.. @j0sien .. when you are trying to give the filename of the output CSV file.. you have to use -o in front of it. So your command will be 
`asreview notes_export ScreeningRound2.asreview -o ScreeningRound2.csv`

**Reply by rohitgarud** [reply]
Hi.. @j0sien .. I have updated the [asreview-notes-export](https://github.com/rohitgarud/asreview-notes-export) after addressing the issue you faced.. please check and let me know if it works as expected..  you can run `pip install --upgrade asreview-notes-export` to get the updated version.. Thank you again for your feedback.

**Reply by j0sien** [reply]
Thanks @rohitgarud . Unfortunately I still get an error after updating:

C:\Users\josien.boetje\OneDrive - Stichting Hogeschool Utrecht\PhD\Study 1 - Systematic Literature Review\d2_Results>asreview notes_export ScreeningRound2.asreview -o ScreeningRound2.csv
Traceback (most recent call last):
  File "c:\program files\python38\lib\runpy.py", line 193, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "c:\program files\python38\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\josien.boetje\AppData\Roaming\Python\Python38\Scripts\asreview.exe\__main__.py", line 7, in <module>
  File "C:\Users\josien.boetje\AppData\Roaming\Python\Python38\site-packages\asreview\__main__.py", line 78, in main
    entry.load()().execute(sys.argv[2:])
  File "C:\Users\josien.boetje\AppData\Roaming\Python\Python38\site-packages\asreviewcontrib\notes_export\entrypoint.py", line 74, in execute
    export_notes(
  File "C:\Users\josien.boetje\AppData\Roaming\Python\Python38\site-packages\asreviewcontrib\notes_export\notes_export.py", line 47, in export_notes
    screening = int(col.split("_")[2])
IndexError: list index out of range

C:\Users\josien.boetje\OneDrive - Stichting Hogeschool Utrecht\PhD\Study 1 - Systematic Literature Review\d2_Results>

Also when trying from a non-cloud folder:
C:\Users\josien.boetje>asreview notes_export ScreeningRound2.asreview -o ScreeningRound2.csv
Traceback (most recent call last):
  File "c:\program files\python38\lib\runpy.py", line 193, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "c:\program files\python38\lib\runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "C:\Users\josien.boetje\AppData\Roaming\Python\Python38\Scripts\asreview.exe\__main__.py", line 7, in <module>
  File "C:\Users\josien.boetje\AppData\Roaming\Python\Python38\site-packages\asreview\__main__.py", line 78, in main
    entry.load()().execute(sys.argv[2:])
  File "C:\Users\josien.boetje\AppData\Roaming\Python\Python38\site-packages\asreviewcontrib\notes_export\entrypoint.py", line 74, in execute
    export_notes(
  File "C:\Users\josien.boetje\AppData\Roaming\Python\Python38\site-packages\asreviewcontrib\notes_export\notes_export.py", line 47, in export_notes
    screening = int(col.split("_")[2])
IndexError: list index out of range

C:\Users\josien.boetje>

**Reply by j0sien** [reply]
I do see a folder 'tmp_data_notes_export'

**Reply by rohitgarud** [reply]
Oh.. I think I missed one scenario while testing my code.. sorry for that.. will update as soon as possible and get back to you..

**Reply by rohitgarud** [reply]
I have updated the version again.. I hope it works this time.. please check and let me know.. thank you.. @j0sien

**Comment by cosalofa** [comment]
Hi @rohitgarud, thanks for creating this extension. It worked for me the first time but I tried it again today and received an error. Please see below.

> C:\Users\Documents\travail_sauvegarde\Meta-analysis\ASReview>asreview notes_export biodiv-3-3.asreview -o biodiv3as.csv

> 
> Traceback (most recent call last):
>   File "<frozen runpy>", line 198, in _run_module_as_main
>   File "<frozen runpy>", line 88, in _run_code
>   File "C:\Users\AppData\Local\Programs\Python\Python311\Scripts\asreview.exe\__main__.py", line 7, in <module>
>   File "C:\Users\AppData\Local\Programs\Python\Python311\Lib\site-packages\asreview\__main__.py", line 78, in main
>     entry.load()().execute(sys.argv[2:])
>   File "C:\Users\AppData\Local\Programs\Python\Python311\Lib\site-packages\asreviewcontrib\notes_export\entrypoint.py", line 63, in execute
>     export_notes(
>   File "C:\Users\AppData\Local\Programs\Python\Python311\Lib\site-packages\asreviewcontrib\notes_export\notes_export.py", line 36, in export_notes
>     dataset_with_results = dataset.df.join(df.set_index('record_id')[['labeling_order','label','exported_notes']], on='record_id')
>                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandas\core\frame.py", line 9976, in join
>     return self._join_compat(
>            ^^^^^^^^^^^^^^^^^^
>   File "C:\Users\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandas\core\frame.py", line 10015, in _join_compat
>     return merge(
>            ^^^^^^
>   File "C:\Users\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandas\core\reshape\merge.py", line 124, in merge
>     return op.get_result(copy=copy)
>            ^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandas\core\reshape\merge.py", line 775, in get_result
>     result = self._reindex_and_concat(
>              ^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandas\core\reshape\merge.py", line 729, in _reindex_and_concat
>     llabels, rlabels = _items_overlap_with_suffix(
>                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "C:\Users\AppData\Local\Programs\Python\Python311\Lib\site-packages\pandas\core\reshape\merge.py", line 2458, in _items_overlap_with_suffix
>     raise ValueError(f"columns overlap but no suffix specified: {to_rename}")
> 
> ValueError: columns overlap but no suffix specified: Index(['labeling_order'], dtype='object')

I think that might be getting this error because I uploaded data from a previous project with new entries to a new project. There was already a column called "labelling_order" in the uploaded file. I tried deleting that column in the temp files (C:\Users\AppData\Local\Temp\tmp71e0q6ax\data) but I still get the error.

Any help with this problem will be greatly appreciated.

**Reply by rohitgarud** [reply]
Hi @cosalofa, thank you for reporting the issue. May I know whether the `exported_notes` column is also present in the dataset? Can you also please check the version of asreview-notes-export using  `asreview notes-export --version`. If it is 0.0.3, please upgrade using `pip install --upgrade asreview-notes-export`, if it is 0.0.5, please let me know.

**Reply by cosalofa** [reply]
Thank you for your quick response @rohitgarud. I had the 0.0.3 version which I have now upgraded. I have a column entitled `exported_notes1` in the dataset. I have rerun the command line and I now get a different error:

> C:\Users\Documents\travail_sauvegarde\Meta-analysis\ASReview>asreview notes_export biodiv-3-3.asreview
> Traceback (most recent call last):
>   File "<frozen runpy>", line 198, in _run_module_as_main
>   File "<frozen runpy>", line 88, in _run_code
>   File "C:\Users\AppData\Local\Programs\Python\Python311\Scripts\asreview.exe\__main__.py", line 7, in <module>
>   File "C:\Users\AppData\Local\Programs\Python\Python311\Lib\site-packages\asreview\__main__.py", line 78, in main
>     entry.load()().execute(sys.argv[2:])
>   File "C:\Users\AppData\Local\Programs\Python\Python311\Lib\site-packages\asreviewcontrib\notes_export\entrypoint.py", line 74, in execute
>     export_notes(
>   File "C:\Users\AppData\Local\Programs\Python\Python311\Lib\site-packages\asreviewcontrib\notes_export\notes_export.py", line 49, in export_notes
>     screening = int(col.split("_")[2])
>                     ~~~~~~~~~~~~~~^^^
> IndexError: list index out of range

Any idea what the issue might be?

**Reply by rohitgarud** [reply]
Thank you. I think I know what is the issue. Did you manually rename the exported_notes column at some point to `exported_notes1` or it was there when you exported the first time?

**Comment by cosalofa** [comment]
I do not remember but it sounds like something that I might have done. I think that I renamed some columns because i had trouble uploading the dataset due to duplicate column names.

**Reply by rohitgarud** [reply]
Let me update the code to address your issue. I will let you know once I update the PyPI version so that you can pip upgrade.

**Reply by rohitgarud** [reply]
@cosalofa I have updated the Pypi version, please upgrade using `pip install --upgrade asreview-notes-export` and let me know. The version should be 0.0.6

**Reply by cosalofa** [reply]
It works perfectly now. Thanks for your help and for creating this very useful extension!

**Comment by Rensvandeschoot** [comment]
resolved in https://github.com/asreview/asreview/pull/1335
